llvm/lib/Transforms/Scalar/IndVarSimplify.cpp

```diff
@@ -995,8 +999,10 @@ linearFunctionTestReplace(Loop *L, BasicBlock *ExitingBB,
       BO->setHasNoSignedWrap(AR->hasNoSignedWrap());
   }
 
-  Value *ExitCnt = genLoopLimit(
-      IndVar, ExitingBB, ExitCount, UsePostInc, L, Rewriter, SE);
+  Value *ExitCnt = genLoopLimit(IndVar, ExitingBB, ExitCount, UsePostInc, L,
+                                Rewriter, SE, TTI);
+  if (!ExitCnt)
+    return false;
```

Comment:
This check should happen earlier. I think you'll want to fold this into the high cost expansion check before linearFunctionTestReplace gets called. It currently works on ExitCount, but it should work on the value that actually gets expanded.